iothub/service/src/IotHubCredential.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Azure.Amqp;
+using Microsoft.Azure.Devices.Common;
+
+namespace Microsoft.Azure.Devices
+{
+    /// <summary>
+    /// The properties required for authentication to IoT hub that are independent of the authentication type.
+    /// </summary>
+    internal abstract class IotHubCredential
+        : IAuthorizationHeaderProvider, ICbsTokenProvider
+    {
+        private const string HostNameSeparator = ".";
+        private const string HttpsEndpointPrefix = "https";
+
+        // Azure.Core (used in IotHubTokenCredential) is not available in NET451.
+        // So we need this constructor for the build to pass.
+        protected IotHubCredential()
+        {
+        }
+
+        protected IotHubCredential(string hostName)
+        {
+            HostName = hostName;
+            IotHubName = GetIotHubName(hostName);
+            AmqpEndpoint = new UriBuilder(CommonConstants.AmqpsScheme, HostName, AmqpConstants.DefaultSecurePort).Uri;
+            HttpsEndpoint = new UriBuilder(HttpsEndpointPrefix, HostName).Uri;
+        }
+
+        public string IotHubName { get; protected set; }
+
+        public string HostName { get; protected set; }
+
+        public Uri HttpsEndpoint { get; protected set; }
+
+        public Uri AmqpEndpoint { get; protected set; }
+
+        public abstract string GetAuthorizationHeader();
+
+        public abstract Task<CbsToken> GetTokenAsync(Uri namespaceAddress, string appliesTo, string[] requiredClaims);
+
+        public Uri BuildLinkAddress(string path)
+        {
+            var builder = new UriBuilder(AmqpEndpoint)
+            {
+                Path = path,
+            };
+
+            return builder.Uri;
+        }
+
+        private static string GetIotHubName(string hostName)
```

Comment:
nit: Can we think of a way to unit test this method? Because it's private, we might have to change the access modifier. 

String parsing logic is something that I try to add coverage for because input can vary greatly, especially if 'hostName' is something passed from the user.

Reply:
Sure let me take a look. I'll add it as a task and address it in my next PR. The class is internal so I think it should be doable. But the test won't really test based on what user passes. The test can only test if the user sends in the correct hostname, we parse it correctly. There is only one format for the hostname.